### PR TITLE
Investigate wildcard-based type parameter reduction

### PR DESCRIPTION
**Pull request by laughedelic**

### CODE REVIEW

File: src/main/java/com/bio4j/angulillos/TypedElement.java [hunk 1]
```diff
@@ -14,10 +14,10 @@
   `E` refers to the element itself, and `ET` its type. You cannot define one without defining the other.
 */
 interface TypedElement <
-  E extends TypedElement<E,ET,G,I,RV,RE>,
-  ET extends TypedElement.Type<E,ET,G,I,RV,RE>,
-  G extends TypedGraph<G,I,RV,RE>,
-  I extends UntypedGraph<RV,RE>, RV,RE
+  E extends TypedElement<E,ET,G,ER>,
+  ET extends TypedElement.Type,
```
Comment (eparejatobes) [inline src/main/java/com/bio4j/angulillos/TypedElement.java]: this totally misses the point of mutually recursive params

File: src/main/java/com/bio4j/angulillos/Property.java [hunk 2]
```diff
@@ -6,17 +6,13 @@
   A property of the [Element](TypedElement.java.md) `N`, with value type `V`.
 */
 interface Property <
-  // the element type
-  N extends TypedElement<N,NT,G,I,RV,RE>, NT extends TypedElement.Type<N,NT,G,I,RV,RE>,
-  // the property type and its value type
-  P extends Property<N,NT,P,V,G,I,RV,RE>, V,
-  // graph
-  G extends TypedGraph<G,I,RV,RE>, I extends UntypedGraph<RV,RE>, RV,RE
+  ET extends TypedElement.Type,
```
Comment (eparejatobes) [inline src/main/java/com/bio4j/angulillos/Property.java]: fine modulo keeping the TypedElement.Type as `NT extends TypedElement.Type<?,NT,?,?,?,?>`

**Comment by laughedelic** [comment]
Beware, you may not like what you will see

File: src/main/java/com/bio4j/angulillos/Property.java (hunk 2 above)
Comment (laughedelic) [inline src/main/java/com/bio4j/angulillos/Property.java]: What's the point in adding parameters to `TypedElement.Type` if it doesn't need any?

**Comment by eparejatobes** [comment]
In general elements and their types should have the same typing context

File: src/main/java/com/bio4j/angulillos/Property.java (hunk 2 above)
Comment (eparejatobes) [inline src/main/java/com/bio4j/angulillos/Property.java]: It needs them. As I said below, XYZ and XYZ.Type need the same typing context

**Comment by eparejatobes** [comment]
How I see this

1. UntypedGraph could be left without F-bounded stuff *if* we assume that it is only going to be used from a typed graph
2. Property could be reduced to just NT and V, filling the rest with wildcards
3. In *method signatures* we can safely substitute for wildcards those types which are left unused, and are only there for filling holes of other types

**Comment by laughedelic** [comment]
Could you elaborate on why

> elements and their types should have the same typing context

please?

**Comment by eparejatobes** [comment]
The idea is to inextricably link elements and their types, and you can only do that with mutually rec F bounded polymorphic signatures, which in turn forces them to share generic type parameters.

**Comment by laughedelic** [comment]
I think that in general it's not needed, if they could just reference to each other, it would be enough (another thing that in the current code it's not possible, but vertex/edge type, for example, doesn't needs to refer to itself).

But ok, I'm fine with the idea of mutual reference: `Vertex<V,VT, ...>` and `Vertex.Type<V,VT, ...>`, but if Vertex needs a reference to the graph and therefore will have a parameter `<G extends TypedGraph...>`, what's the point in adding it to the `Vertex.Type`? The type doesn't use graph and doesn't need that type.

**Comment by eparejatobes** [comment]
I don't see much gain in removing the graph param, and different parameter lists is a tad confusing. No big deal though.

**Comment by eparejatobes** [comment]
We need to allow edges between vertices coming from different graphs, lest we lose the ability to compose graphs

File: src/main/java/com/bio4j/angulillos/TypedElement.java [hunk 3]
```diff
@@ -44,7 +44,13 @@
...
     Element types are also used as factories for constructing instances of the corresponding elements.
   */
-  interface Type {
+  interface Type <
+    E extends TypedElement<E,ET,G,ER>,
+    ET extends TypedElement.Type<E,ET,G,ER>,
+    G extends TypedGraph<G,?,?>,
+    ER
+  >
+  {
```
Comment (laughedelic) [inline src/main/java/com/bio4j/angulillos/TypedElement.java]: Can you tell me, where do you use `TypedElement.Type` with all those parameters? I don't see any place. It's just a base trait to add this `name()` to both vertex type and edge type.

File: src/main/java/com/bio4j/angulillos/TypedEdge.java [hunk 4]
```diff
@@ -53,18 +54,19 @@
...
-    T extends TypedVertex<T,TT,TG,?>,
-    TT extends TypedVertex.Type<T,TT,TG,?>,
-    TG extends TypedGraph<TG,?,?>,
+    T extends TypedVertex<T,TT,TG,RV,RE>,
+    TT extends TypedVertex.Type<T,TT,TG,RV,RE>,
+    TG extends TypedGraph<TG,RV,RE>,
     // raw
+    RV,
     RE
   > extends
```
Comment (laughedelic) [inline src/main/java/com/bio4j/angulillos/TypedEdge.java]: Both in `TypedEdge` and `TypedEdge.Type`: what is the point of having explicit references to `SG` and `TG` graphs?

File: src/main/java/com/bio4j/angulillos/TypedGraph.java [hunk 5]
```diff
@@ -31,10 +31,10 @@ V addVertex(TypedVertex.Type<V,?,G,RV> vertexType) {
 
   /* adds an edge; note that this method does not set any properties. As it needs to be called by vertices in possibly different graphs, all the graph bounds are free with respect to G. */
   default <
-    S extends TypedVertex<S,?,? extends TypedGraph<?,RV,RE>,RV>,
-    R extends TypedEdge<S,?,?, R,RT,G, T,?,?, RE>,
-    RT extends TypedEdge.Type<S,?,?, R,RT,G, T,?,?, RE>,
-    T extends TypedVertex<T,?,? extends TypedGraph<?,RV,RE>,RV>
+    S extends TypedVertex<S,?,? extends TypedGraph<?,RV,RE>,RV,RE>,
+    R extends TypedEdge<S,?,? extends TypedGraph<?,RV,RE>, R,RT,? extends TypedGraph<?,RV,RE>, T,?,? extends TypedGraph<?,RV,RE>, RV,RE>,
+    RT extends TypedEdge.Type<S,?,? extends TypedGraph<?,RV,RE>, R,RT,? extends TypedGraph<?,RV,RE>, T,?,? extends TypedGraph<?,RV,RE>, RV,RE>,
```
Comment (laughedelic) [inline src/main/java/com/bio4j/angulillos/TypedGraph.java]: Now _this_ is something interesting :+1:

File: src/main/java/com/bio4j/angulillos/TypedVertex.java [hunk 6]
```diff
@@ -10,27 +10,31 @@
...
 interface TypedVertex <
-  N extends TypedVertex<N,NT,G,RV>,
-  NT extends TypedVertex.Type<N,NT,G,RV>,
-  G extends TypedGraph<G,RV,?>,
-  RV
+  N   extends TypedVertex<N,NT,G,RV,RE>,
+  NT  extends TypedVertex.Type<N,NT,G,RV,RE>,
+  G   extends TypedGraph<G,RV,RE>,
+  RV,
+  RE // NOTE we need RE if we want to safely add edges
```
Comment (laughedelic) [inline src/main/java/com/bio4j/angulillos/TypedVertex.java]: I don't want. I think that having typesafe `addVertex`, `addEdge` in the `TypedGraph` is more than enough. They are completely non-essential here and this parameter is needed _only_ for them.

Comment (eparejatobes) [inline src/main/java/com/bio4j/angulillos/TypedVertex.java]: but then you cannot call it from here :P

**Comment by laughedelic** [comment]
What we've decided here:

- `TypedX` and `TypedX.Type` to have same parameters lists
- `Property<ET,V>`
- `TypedElement<E,ET,G,ERaw>`
- `TypedVertex`: remove `RE` (then it has the same parameters as `TypedElement`, but specialized)
- `TypedEdge`: remove `SG` and `TG`, keep `G,RV,RE`

**Comment by laughedelic** [comment]
What's left here:
- [ ] restore index-related code
- [ ] restore or remove (for now) vertex-query code
- [ ] update twitter-graph example

**Comment by laughedelic** [comment]
Hey @eparejatobes! Check out the twitter schema code here. I tried to reduce the boilerplate in schema definition as much as possible and I think that it's quite nice now.

Also, I just realised, that we can use this trick to forget about overriding `self()` everywhere: 

```java
// in the `TypedElement` interface
default F self() { return type().fromRaw( raw() ); }
```

But I guess, that it will lead to creating new objects on each call. What do you think?

**Comment by laughedelic** [comment]
Also, I think this is ready to merge :shipit: Quite a big refactoring PR, but I think it's also a big improvement :+1: :tada:

@eparejatobes review it, please.

**Comment by laughedelic** [comment]
The 4cf16da commit is just a piece of cake! :cake: :tada: 
¡zero-boilerpate in the schema definition: **achieved**!

**Pull request merged**